File: BOJ/1654-랜선 자르기/solution.cpp
#include <bits/stdc++.h>
using namespace std;
int k, n;
vector<long long> lan;

void input(){
    cin >> k >> n;

    lan.resize(k);
    for(int i = 0; i < k; i++) cin >> lan[i];
    sort(lan.begin(), lan.end());
}

long long count_lans(long long length){
    long long res = 0;
    for(int i = 0; i < k; i++){
        res += (lan[i] / length);
    }
    return res;
}

void solve(){
    long long left = 1;
    long long right = LLONG_MAX - 1;
    while(left <= right){
        long long mid = (left + right) / 2;

        long long here = count_lans(mid);
        if(here < n){
            right = mid - 1;
        }
        else{
            left = mid + 1;
        }
    }
    cout << right << "\n";
}

int main(void){
    cin.tie(0), ios::sync_with_stdio(0);
    input();
    solve();
    return 0;
}

File: BOJ/1654-랜선 자르기/solution.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
)

var k, n int
var list []int
var reader = bufio.NewReader(os.Stdin)
var writer = bufio.NewWriter(os.Stdout)

func input() {
	fmt.Fscan(reader, &k, &n)

	var lan int
	for i := 0; i < k; i++ {
		fmt.Fscan(reader, &lan)
		list = append(list, lan)
	}
}

func splitLan(length int) int {
	res := 0
	for i := 0; i < k; i++ {
		res += (list[i] / length)
	}
	return res
}

func solve() {
	sort.Ints(list)

	left := 1
	right := list[len(list)-1]
	for left <= right {
		mid := (left + right) / 2
		check := splitLan(mid)

		if check >= n {
			left = mid + 1
		} else {
			right = mid - 1
		}
	}
	fmt.Fprintln(writer, right)
}

func main() {
	defer writer.Flush()
	input()
	solve()
}
